base/channels.jl

```diff
@@ -472,3 +472,29 @@ function iterate(c::Channel, state=nothing)
 end
 
 IteratorSize(::Type{<:Channel}) = SizeUnknown()
+
+"""
+    Threads.foreach(f, channel::Channel; ntasks=Threads.nthreads())
+
+Similar to `foreach(f, channel)`, but iteration over `channel` and calls to
+`f` are split across `ntasks` Tasks spawned by `Threads.@spawn`. This function
+will wait for all internally spawned tasks to complete before returning.
+"""
+function Threads.foreach(f, channel::Channel; ntasks=Threads.nthreads())
+    stop = Threads.Atomic{Bool}(false)
+    @sync for _ in 1:ntasks
+        Threads.@spawn try
+            for item in channel
+                _waitspawn(f, item)
+                stop[] && break  # do this _after_ f(item) to avoid loosing `item`
```

Comment:
Hmm... Actually, this is not enough because a task could be stuck on `take!` at `for item in channel`.  But I'm not sure how to terminate the tasks without closing the `channel` or loosing an `item`.  I think we need builtin "`select`" (as in Go) for this.